03:17:06 [ERROR] client socket_err: 127.0.0.1:48304 failed to fill whole buffer

Hi there, I was trying out your crate today and noticed this error: `03:17:06 [ERROR] client socket_err: 127.0.0.1:48304 failed to fill whole buffer`.
  
I spent some time trying to fix it, and I am new to rust and ISO 8583, so maybe you can help me solve the problem and learn something along the way.

When I run `test_send_recv_iso_1100` the server emits this error, but it may be misleading.  `server.mli.parse` Throws an IsoError with `std::io::error::ErrorKind::UnexpectedEof` [here](https://github.com/rkbalgi/iso8583_rs/blob/0bc7ede98c002f2acc7dfc38394c1a45bd221b8b/src/iso8583/mli.rs#L40).

It may be because `in_buff` is being processed twice for some reason?  I can't quite tell.

Hi Thomas, I'm new to Rust too :-) and so yes, we can all learn something.

The error is definitely misleading and I'l try to fix it. What the error is saying really is "the client closed the connection". You can see the same on my github page too :-) [here- scroll to the very end](https://github.com/rkbalgi/iso8583_rs#run-iso-server)

```
15:54:47 [DEBUG] (3) iso8583_rs::iso8583::server: request processing time = 9 millis
15:54:47 [ERROR] client socket_err: 127.0.0.1:56307 failed to fill whole buffer
```

(So far I have been able to make out that in rust if a socket read returns 0 means that the client has closed the connection. In other languages you might get a -1 or a SocketClosedException)

BTW, when you run the server (_main.rs_) and then the _test_send_recv_iso_1100_, do you see the server processing the request and the test side receiving the response?

I've made some changes on the master branch to report a slightly meaningful message -

```
17:48:28 [DEBUG] (3) iso8583_rs::iso8583::server: request processing time = 2 millis
17:48:28 [ERROR] client socket_err: 127.0.0.1:50920 connection closed. cause: UnexpectedEof
```

> Hi Thomas, I'm new to Rust too :-) and so yes, we can all learn something.
> 
> The error is definitely misleading and I'l try to fix it. What the error is saying really is "the client closed the connection". You can see the same on my github page too :-) [here- scroll to the very end](https://github.com/rkbalgi/iso8583_rs#run-iso-server)
> 
> ```
> 15:54:47 [DEBUG] (3) iso8583_rs::iso8583::server: request processing time = 9 millis
> 15:54:47 [ERROR] client socket_err: 127.0.0.1:56307 failed to fill whole buffer
> ```
> 
> (So far I have been able to make out that in rust if a socket read returns 0 means that the client has closed the connection. In other languages you might get a -1 or a SocketClosedException)
> 
> BTW, when you run the server (_main.rs_) and then the _test_send_recv_iso_1100_, do you see the server processing the request and the test side receiving the response?

Yes, I see the test client receive a response, then it's like it sends another response back to the server.

--- a/src/iso8583/mli.rs
+++ b/src/iso8583/mli.rs
@@ -2,6 +2,8 @@
 use crate::iso8583::IsoError;
 use byteorder::{WriteBytesExt, ReadBytesExt};
 use std::io::{Read, ErrorKind, Error};
+use std::net::{TcpStream};
+
 
 
 pub enum MLIType {
@@ -16,6 +18,8 @@ pub trait MLI: Sync + Send {
     fn parse(&self, in_buf: &mut dyn Read) -> Result<u32, IsoError>;
     /// Creates a Vec<u8> that represents the MLI containing n bytes
     fn create(&self, n: &usize) -> Result<Vec<u8>, IsoError>;
+    /// Checks to see if data is available for `MLI::parse`
+    fn is_available(&self, stream: &TcpStream) -> Result<bool, IsoError>;
 }
 
 /// This struct represents an MLI of 2E (i.e 2 bytes of length indicator exclusive of its own length)

--- a/src/iso8583/mli.rs
+++ b/src/iso8583/mli.rs
@@ -67,9 +67,15 @@ impl MLI for MLI2E {
     fn is_available(&self, stream: &TcpStream) -> Result<bool, IsoError> {
         let mut buf = vec![0; 2];
 
-        match stream.peek(&mut buf).unwrap() {
-            2 => Ok(true),
-            _ => Err(IsoError { msg: format!("Client disconnected")})
+        match stream.peek(&mut buf) {
+            Ok(n) => {
+                if n == 2 {
+                    Ok(true)
+                } else {
+                    Err(IsoError { msg: format!("client disconnected") })
+                }
+            }
+            Err(e) => Err(IsoError { msg: format!("stream err. cause: {}", e.to_string()) })
         }
     }
 }
@@ -94,9 +100,15 @@ impl MLI for MLI4E {
     fn is_available(&self, stream: &TcpStream) -> Result<bool, IsoError> {
         let mut buf = vec![0; 4];
 
-        match stream.peek(&mut buf).unwrap() {
-            4 => Ok(true),
-            _ => Err(IsoError { msg: format!("Client disconnected")})
+        match stream.peek(&mut buf) {
+            Ok(n) => {
+                if n == 4 {
+                    Ok(true)
+                } else {
+                    Err(IsoError { msg: format!("client disconnected") })
+                }
+            }
+            Err(e) => Err(IsoError { msg: format!("stream err. cause: {}", e.to_string()) })
         }
     }
 }
@@ -121,9 +133,15 @@ impl MLI for MLI2I {
     fn is_available(&self, stream: &TcpStream) -> Result<bool, IsoError> {
         let mut buf = vec![0; 2];
 
-        match stream.peek(&mut buf).unwrap() {
-            2 => Ok(true),
-            _ => Err(IsoError { msg: format!("Client disconnected")})
+        match stream.peek(&mut buf) {
+            Ok(n) => {
+                if n == 2 {
+                    Ok(true)
+                } else {
+                    Err(IsoError { msg: format!("client disconnected") })
+                }
+            }
+            Err(e) => Err(IsoError { msg: format!("stream err. cause: {}", e.to_string()) })
         }
     }
 }
@@ -147,9 +165,15 @@ impl MLI for MLI4I {
     fn is_available(&self, stream: &TcpStream) -> Result<bool, IsoError> {
         let mut buf = vec![0; 4];
 
-        match stream.peek(&mut buf).unwrap() {
-            4 => Ok(true),
-            _ => Err(IsoError { msg: format!("Client disconnected")})
+        match stream.peek(&mut buf) {
+            Ok(n) => {
+                if n == 4 {
+                    Ok(true)
+                } else {
+                    Err(IsoError { msg: format!("client disconnected") })
+                }
+            }
+            Err(e) => Err(IsoError { msg: format!("stream err. cause: {}", e.to_string()) })
         }
     }
 }

--- a/src/iso8583/mli.rs
+++ b/src/iso8583/mli.rs
@@ -63,6 +63,15 @@ impl MLI for MLI2E {
         let _ = mli.write_u16::<byteorder::BigEndian>(n.clone() as u16);
         Ok(mli)
     }
+
+    fn is_available(&self, stream: &TcpStream) -> Result<bool, IsoError> {
+        let mut buf = vec![0; 2];
+
+        match stream.peek(&mut buf).unwrap() {
+            2 => Ok(true),
+            _ => Err(IsoError { msg: format!("Client disconnected")})
+        }
+    }
 }
 
 
@@ -80,6 +89,15 @@ impl MLI for MLI4E {
         let mut mli = Vec::<u8>::new();
         let _ = mli.write_u32::<byteorder::BigEndian>(n.clone() as u32);
         Ok(mli)
+    }
+
+    fn is_available(&self, stream: &TcpStream) -> Result<bool, IsoError> {
+        let mut buf = vec![0; 4];
+
+        match stream.peek(&mut buf).unwrap() {
+            4 => Ok(true),
+            _ => Err(IsoError { msg: format!("Client disconnected")})
+        }
     }
 }
 
@@ -99,6 +117,15 @@ impl MLI for MLI2I {
         let _ = mli.write_u16::<byteorder::BigEndian>((n.clone() as u16) + 2);
         Ok(mli)
     }
+
+    fn is_available(&self, stream: &TcpStream) -> Result<bool, IsoError> {
+        let mut buf = vec![0; 2];
+
+        match stream.peek(&mut buf).unwrap() {
+            2 => Ok(true),
+            _ => Err(IsoError { msg: format!("Client disconnected")})
+        }
+    }
 }
 
 impl MLI for MLI4I {
@@ -115,6 +142,15 @@ impl MLI for MLI4I {
         let mut mli = Vec::<u8>::new();
         let _ = mli.write_u32::<byteorder::BigEndian>((n.clone() as u32) + 4);
         Ok(mli)
+    }
+
+    fn is_available(&self, stream: &TcpStream) -> Result<bool, IsoError> {
+        let mut buf = vec![0; 4];
+
+        match stream.peek(&mut buf).unwrap() {
+            4 => Ok(true),
+            _ => Err(IsoError { msg: format!("Client disconnected")})
+        }
     }
 }
 

--- a/src/iso8583/server.rs
+++ b/src/iso8583/server.rs
@@ -110,7 +110,8 @@ fn new_client(iso_server: &ISOServer, stream_: TcpStream) {
             if reading_mli {
                 match server.mli.is_available(&stream) {
                     Ok(true) => true,
-                    IsoError => break 'done,
+                    Ok(false) => false,
+                    Err(e) => break 'done,
                 };
 
                 match server.mli.parse(&mut reader) {

--- a/src/iso8583/server.rs
+++ b/src/iso8583/server.rs
@@ -108,13 +108,10 @@ fn new_client(iso_server: &ISOServer, stream_: TcpStream) {
         'done:
         loop {
             if reading_mli {
-                let mut peek_buf = vec![0; 1];
-                let bytes_read = stream.peek(&mut peek_buf).unwrap();
-
-                if bytes_read == 0 {
-                    println!("client closed the connection.");
-                    break 'done;
-                }
+                match server.mli.is_available(&stream) {
+                    Ok(true) => true,
+                    IsoError => break 'done,
+                };
 
                 match server.mli.parse(&mut reader) {
                     Ok(n) => {

--- a/src/iso8583/server.rs
+++ b/src/iso8583/server.rs
@@ -108,6 +108,14 @@ fn new_client(iso_server: &ISOServer, stream_: TcpStream) {
         'done:
         loop {
             if reading_mli {
+                let mut peek_buf = vec![0; 2];
+                let bytes_read = stream.peek(&mut peek_buf).unwrap();
+
+                if bytes_read == 0 {
+                    println!("client closed the connection.");
+                    break 'done;
+                }
+
                 match server.mli.parse(&mut reader) {
                     Ok(n) => {
                         mli = n;

--- a/src/iso8583/server.rs
+++ b/src/iso8583/server.rs
@@ -108,7 +108,7 @@ fn new_client(iso_server: &ISOServer, stream_: TcpStream) {
         'done:
         loop {
             if reading_mli {
-                let mut peek_buf = vec![0; 2];
+                let mut peek_buf = vec![0; 1];
                 let bytes_read = stream.peek(&mut peek_buf).unwrap();
 
                 if bytes_read == 0 {

